service/debugger/debugger.go

```diff
@@ -168,16 +178,66 @@ func (d *Debugger) CreateBreakpoint(requestedBp *api.Breakpoint) (*api.Breakpoin
 	if err != nil {
 		return nil, err
 	}
-	bp.Tracepoint = requestedBp.Tracepoint
-	bp.Goroutine = requestedBp.Goroutine
-	bp.Stacktrace = requestedBp.Stacktrace
-	bp.Variables = requestedBp.Variables
-	bp.Cond = nil
+	if err := copyBreakpointInfo(bp, requestedBp); err != nil {
+		d.process.ClearBreakpoint(bp.Addr)
+		return nil, err
+	}
 	createdBp = api.ConvertBreakpoint(bp)
 	log.Printf("created breakpoint: %#v", createdBp)
 	return createdBp, nil
 }
 
+func (d *Debugger) AmendBreakpoint(amend *api.Breakpoint) error {
+	var original *proc.Breakpoint
+	for i := range d.process.Breakpoints {
+		if d.process.Breakpoints[i].ID == amend.ID {
+			original = d.process.Breakpoints[i]
+			break
+		}
+	}
+	if original == nil {
+		return fmt.Errorf("no breakpoint with ID %d", amend.ID)
+	}
+	if err := d.validBreakpointName(amend.Name, false); err != nil {
+		return err
+	}
+	copyBreakpointInfo(original, amend)
+	return nil
+}
+
+func copyBreakpointInfo(bp *proc.Breakpoint, requested *api.Breakpoint) (err error) {
+	bp.Name = requested.Name
+	bp.Tracepoint = requested.Tracepoint
+	bp.Goroutine = requested.Goroutine
+	bp.Stacktrace = requested.Stacktrace
+	bp.Variables = requested.Variables
+	bp.Cond = nil
+	if requested.Cond != "" {
+		bp.Cond, err = parser.ParseExpr(requested.Cond)
+	}
+	return err
+}
+
+func (d *Debugger) validBreakpointName(name string, checkdup bool) error {
```

Comment:
Let's make this a pure function and move the duplicate name check outside of this func.